### PR TITLE
WEB: Better management of releases

### LINKED ISSUE #50885: WEB: Better management of releases in the pandas home page
In the [pandas home page](https://pandas.pydata.org/) we show the released versions, both in the `Latest version` box, and the `Previous versions` list.

Obtaining the releases is done [here](https://github.com/pandas-dev/pandas/blob/main/web/pandas_web.py#L206). So far, the function is pretty simple, just sorting the releases by date, and assuming the last one by date it's the latest.

While this makes sense, we're planning to release 2.0 shortly, and it's likely that we release 1.5.4 after it. With the current implementation and after those releases, we'd list 1.5.4 as the latest release, which is not great since we expect users to use 2.0, which is not the latest by date, but the newest in terms of features.

The idea is to improve the current implementation to be smarter, and not use the date, but use the version number. This can be done with [version.parse](https://packaging.pypa.io/en/latest/version.html#packaging.version.parse).

Since we are implementing this, I think besides detecting the latest version, it can make sense to also filter obsolete versions from the long list of releases. For example, 1.5.3 is the same version as 1.5.2, 1.5.1 and 1.5.0 but with bugs fixed. There is no reason a user would want to install any 1.5 version except the last one. 1.4.4 is a reasonable version, since users may have code using that version that they didn't migrate yet to be compatible with 1.5. But 1.4.3, 1.4.2, 1.4.1 and 1.4.0 shouldn't be used, so probably better to not have them in the list.

I don't think we have any tests for the script that generates the website. But since the new implementation is not trivial, it'd be good to have it tested. Maybe just a `tests/test_pandas_web.py` in the same directory as the script, and calling pytest on that directory in the documentation build is the best option.

### PR DESCRIPTION
**Pull request by Dukastlik**
Added releases sorting by version number and sorting out obsolete releases in web/pandas_web.py

- [ ] closes #50885 
- [x] Tests added and passed
- [x] All code checks passed

### CODE REVIEW

**Comment by Dukastlik** [comment]
Hi @datapythonista !
Added this PR and tests, but was not sure what you've meant in original issue by:
> Maybe just a tests/test_pandas_web.py in the same directory as the script, and calling pytest on that directory in the documentation build is the best option.
Could not figure out how new tests should be called.
Also added `mock_response` fixture that can be used for testing other web features.

File: web/tests/test_pandas_web.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        "published_at": "2023-04-30T13:24:32Z",
+        "tag_name": "v2.0.0",
+        "assets": None,
+    },
+    {
+        "prerelease": True,
+        "published_at": "2023-01-19T03:34:05Z",
+        "tag_name": "v1.5.3xd",
+        "assets": None,
+    },
```
Comment (datapythonista) [inline web/tests/test_pandas_web.py:72]: Maybe we can add 10.0 to make sure the possible bug commented above is not there.

**Review by datapythonista — COMMENTED** [review]
Great job, just one comment

File: web/pandas_web.py [hunk 2]
```diff
@@ -249,7 +251,16 @@ def home_add_releases(context):
...
+        # sorting out obsolete versions
+        grouped_releases = itertools.groupby(
+            context["releases"], key=lambda r: (r["name"].major, r["name"].minor)
+        )
+        context["releases"] = [
+            max(release_group, key=lambda r: r["name"].minor)
+            for _, release_group in grouped_releases
+        ]
+        # sorting releases by version number
+        context["releases"].sort(key=lambda r: r["name"], reverse=True)
```
[the remark this answers is not in the document]
Comment (Dukastlik) [reply web/pandas_web.py:263]: I'm parsing version before that on line 244. Is ok?

**Comment by datapythonista** [comment]
> Could not figure out how new tests should be called. Also added `mock_response` fixture that can be used for testing other web features.

Good point. I think the normal tests just call pytest for the pandas/ directory. In the CI job to build the website and docs we could add a step to call pytest for the tests in web/. I think that's what makes sense, but other ideas are welcome.

File: web/tests/test_pandas_web.py (hunk 1 above)
Comment (Dukastlik) [reply web/tests/test_pandas_web.py:72]: Added

File: web/pandas_web.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dukastlik) [reply web/pandas_web.py]: Done

**Comment by Dukastlik** [comment]
> > Can you please point me out a file with CI job for docs and website? Sorry, I'm new to the project.
> 
> https://github.com/pandas-dev/pandas/blob/main/.github/workflows/docbuild-and-upload.yml

Thanks a lot! Added "Test website" step to ci job.

**Review by mroeschke — APPROVED** [review]
LGTM merge when ready @datapythonista

**Comment by datapythonista** [comment]
Amazing job, thanks for taking care of this @Dukastlik, and sorry that reviews took a while.

**Pull request merged**